File: app/src/main/java/com/kotlin/githubapi/ui/base/paged/BasePagedAdapter.java
package com.kotlin.githubapi.ui.base.paged;

import java.util.Objects;

import androidx.annotation.NonNull;
import androidx.paging.PagedListAdapter;
import androidx.recyclerview.widget.AsyncDifferConfig;
import androidx.recyclerview.widget.DiffUtil;
import androidx.recyclerview.widget.RecyclerView;

public abstract class BasePagedAdapter<T, VH extends RecyclerView.ViewHolder> extends PagedListAdapter<T, VH> {

    public interface OnItemClickListener {

        void onItemClick(int position);
    }

    protected PageState pageState;
    private Boolean isLoading = false;

    public BasePagedAdapter(@NonNull DiffUtil.ItemCallback<T> diffCallback) {
        super(diffCallback);
    }

    public BasePagedAdapter(@NonNull AsyncDifferConfig<T> config) {
        super(config);
    }

    public void setPageState(PageState newPageState) {
        if (getCurrentList() == null || getCurrentList().isEmpty()) return;
        if (Objects.equals(newPageState, PageState.LOADING)) {
            isLoading = true;
            notifyItemChanged(getItemCount() - 1);
        } else if (newPageState != null && newPageState.getPageLoadStatus() != null) {
            switch (newPageState.getPageLoadStatus()) {
                case RUNNING:
                    if (isLoading) return;
                    isLoading = true;
                    notifyItemInserted(getItemCount());
                    break;
                case SUCCESS:
                case EMPTY:
                case FAILED:
                    isLoading = false;
                    notifyItemRemoved(getItemCount());
                    break;
            }
        }
    }

    public Boolean isEmpty() {
        return getItemCount() <= 0;
    }

    public T getItem(@NonNull final Integer position) {
        return super.getItem(position);
    }

    protected boolean hasExtraRow() {
        return isLoading;
    }
}

File: app/src/main/java/com/kotlin/githubapi/ui/base/paged/BasePagedAdapter.kt
package com.kotlin.githubapi.ui.base.paged

import androidx.paging.PagedListAdapter
import androidx.recyclerview.widget.AsyncDifferConfig
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView

abstract class BasePagedAdapter<T : Any, VH : RecyclerView.ViewHolder> : PagedListAdapter<T, VH> {

    protected var pageState: PageState? = null
    private var isLoading: Boolean = false

    val isEmpty: Boolean?
        get() = itemCount <= 0

    constructor(diffCallback: DiffUtil.ItemCallback<T>) : super(diffCallback)

    constructor(config: AsyncDifferConfig<T>) : super(config)

    fun updatePageState(newPageState: PageState) {
        if (currentList == null || currentList!!.isEmpty()) return
        if (newPageState == PageState.LOADING) {
            isLoading = true
            notifyItemChanged(itemCount - 1)
        } else if (newPageState.pageLoadStatus != null) {
            when (newPageState.pageLoadStatus) {
                PageLoadStatus.RUNNING -> {
                    if (isLoading) return
                    isLoading = true
                    notifyItemInserted(itemCount)
                }
                PageLoadStatus.SUCCESS,
                PageLoadStatus.EMPTY,
                PageLoadStatus.FAILED -> {
                    isLoading = false
                    notifyItemRemoved(itemCount)
                }
                else -> error("Please specify all the cases.")
            }
        }
    }

    protected fun hasExtraRow(): Boolean = isLoading
}
